doc/design/logging.md

```diff
@@ -0,0 +1,361 @@
+as log level. To have a different fallback log level, the `logLevelFromEnvOr`
+function can be used, e.g.
+
+```cpp
+iox::log::initLogger(iox::log::logLevelFromEnvOr(iox::log::LogLevel::DEBUG));
+```
+
+If the logger shall not be altered via environment variables, `initLogger` must
+be called with the fitting log level.
+
+For the `TestingLogger` there is an additional environment variable called
+`IOX_TESTING_ALLOW_LOG`. This enables the printing of the log messages for all
+tests instead of only for failed ones.
+
+| Environment variable | Allowed values |
+|--------------|-----------|
+| IOX_LOG_LEVEL | off, fatal, error, warn, info, debug, trace |
+| IOX_TESTING_ALLOW_LOG | on, off |
+
+#### Thread local storage
+
+The `Logger` base class is using thread local storage to provide a buffer for
+each thread. In case thread local storage is not desired, the logger must be
+re-implemented via the platform abstraction.
+
+### Code example
+
+#### Using the default console logger
+
+```cpp
+#include "iceoryx_hoofs/log/logging.hpp"
+
+int main()
+{
+    iox::log::initLogger(iox::log::logLevelFromEnvOr(iox::log::LogLevel::DEBUG));
+
+    IOX_LOG(DEBUG) << "Hello World";
+
+    return 0;
+}
+```
+
+#### Add logging support for custom types
+
+```cpp
+#include "iceoryx_hoofs/log/logging.hpp"
+#include "iceoryx_hoofs/log/logstream.hpp"
+#include <cstdint>
+
+struct MyType
+{
+    uint16_t a {0};
+    uint16_t b {0};
+};
+
+iox::log::LogStream& operator<<(iox::log::LogStream& stream, const MyType& m)
+{
+    stream << "MyType { a = " << m.a << "; b = " << m.b << " }";
+    return stream;
+}
+
+int main()
+{
+    MyType m;
+    IOX_LOG(INFO) << m;
+
+    return 0;
+}
+```
+
+#### Creating a custom logger
+
+```cpp
+#include "iceoryx_hoofs/log/logger.hpp"
+#include "iceoryx_hoofs/log/logging.hpp"
+
+class MyLogger : public iox::log::Logger
+{
+  public:
+    static void init()
+    {
+        static MyLogger myLogger;
+        iox::log::setActiveLogger(&myLogger);
+        iox::log::initLogger(iox::log::logLevelFromEnvOr(iox::log::LogLevel::INFO));
+    }
+
+  private:
+    void setupNewLogMessage(const char*, const int, const char*, iox::log::LogLevel logLevel) override
+    {
+        switch(logLevel) {
+            case iox::log::LogLevel::FATAL:
+                logString("💀: ");
+                break;
+            case iox::log::LogLevel::ERROR:
+                logString("🙈: ");
+                break;
+            case iox::log::LogLevel::WARN:
+                logString("🙀: ");
+                break;
+            case iox::log::LogLevel::INFO:
+                logString("💘: ");
+                break;
+            case iox::log::LogLevel::DEBUG:
+                logString("🐞: ");
+                break;
+            case iox::log::LogLevel::TRACE:
+                logString("🐾: ");
+                break;
+            default:
+                logString("🐔: ");
+        }
+    }
+
+    void flush() override {
+        puts(m_buffer);
```

Comment:
OK, `m_buffer` is inherited through the parent class I presume. The user really needs to know what she can and cannot do with this buffer.

Reply:
You are right. This exposes too much of the implementation. I'll change this to

```cpp
puts(Logger::buffer());
Logger::assumePushed();
```

Reply:
Adjusted as proposed